pkg/ebpf/dns_stats.go

```diff
@@ -75,3 +132,33 @@ func (d *dnsStatKeeper) GetAndResetAllStats() map[dnsKey]dnsStats {
 	d.stats = make(map[dnsKey]dnsStats)
 	return ret
 }
+
+func (d *dnsStatKeeper) removeExpiredStates(earliestTs time.Time) {
+	d.mux.Lock()
+	defer d.mux.Unlock()
+	for k, v := range d.state {
+		if v.Before(earliestTs) {
+			delete(d.state, k)
+			d.deleteCount++
+			stats := d.stats[k.key]
+			stats.timeouts++
+			d.stats[k.key] = stats
+		}
+	}
+
+	if d.deleteCount < d.deleteThreshold {
+		return
+	}
+
+	// golang/go#20135 : maps do not shrink after elements removal (delete)
+	copied := make(map[stateKey]time.Time, len(d.state))
+	for k, v := range d.state {
+		copied[k] = v
+	}
+	d.state = copied
+	d.deleteCount = 0
+}
```

Comment:
Most of the times we deal with this we rebuild the map on an interval, and don't check to see how many elements we've removed. From my reading of the ticket, it's not clear that this will stop the issue from happening.